File: public/dsa/linked_lists/reverse_a_linked_list/recursive/reverse_a_linked_list.py

# defs-start
import sys

sys.setrecursionlimit(10**6)


class ListNode:
    def __init__(self, data=0, next=None):
        self.data = data
        self.next = next
# defs-end


class Solution:
    def reverseLinkedList(self, head: ListNode) -> ListNode:
        if head and head.next:
            last = self.reverseLinkedList(head.next)
            head.next.next = head
            head.next = None
            return last
        return head


# tests-start
head = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
sol = Solution()
res = sol.reverseLinkedList(head)

temp = res
while temp:
    print(temp.data, end=' ')
    temp = temp.next

# tests-end


'''output
4 3 2 1
'''

File: public/dsa/linked_lists/reverse_a_linked_list/recursive/reverse_a_linked_list.cpp

#include <bits/stdc++.h>
using namespace std;

// defs-start
class ListNode {
   public:
    int data;
    ListNode* next;
    ListNode(int data, ListNode* next = nullptr) : data(data), next(next) {}
};

// defs-end

ListNode* reverseLinkedList(ListNode* head) {
    if (head && head->next) {
        ListNode* last = reverseLinkedList(head->next);
        head->next->next = head;
        head->next = nullptr;
        return last;
    }
    return head;
}

// tests-start
int main() {
    ListNode* head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));

    ListNode* res = reverseLinkedList(head);

    ListNode* temp = res;
    while (temp) {
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout << endl;

    return 0;
}
// tests-end

/*output
4 3 2 1
*/